### PR TITLE
[SYCL][ESIMD] BFN function implementation

### PR DESCRIPTION
**Pull request by turinevgeny**
API follows a similar one offered by CM.
Example: d = esimd::bfn<~bfn_t::x & ~bfn_t::y & ~bfn_t::z>(s0, s1, s2);

### CODE REVIEW

File: sycl/include/sycl/ext/intel/experimental/esimd/math.hpp
Comment (sarnex) [inline sycl/include/sycl/ext/intel/experimental/esimd/math.hpp:1748]: is it possible to put all the bfn_t stuff inside a stuct so it's encapsulated? or does it have to be seperate?

also is it possible to move it to some detail header?

File: sycl/include/sycl/ext/intel/experimental/esimd/math.hpp
Comment (sarnex) [inline sycl/include/sycl/ext/intel/experimental/esimd/math.hpp:1783]: just for my own interest where does the 2/4 byte requirement come from? does the underlying hard instruction only support these?

File: sycl/test/esimd/intrins_trans.cpp
Comment (sarnex) [inline sycl/test/esimd/intrins_trans.cpp:310]: sorry also for my own interest, the `i8 -1` argument is from the 0xff template argument right?

**Review by sarnex — COMMENTED** [review]
mostly lgtm, but prolly should get a second review for correctness

File: sycl/include/sycl/ext/intel/experimental/esimd/math.hpp
Comment (v-klochkov) [inline sycl/include/sycl/ext/intel/experimental/esimd/math.hpp:1733]: I may be wrong, but I think this comment should be placed right before the bfn() function definition.
Otherwise, this comment would stick to 'enum class bfn_t'.

File: sycl/include/sycl/ext/intel/experimental/esimd/math.hpp
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/experimental/esimd/math.hpp:1748]: IMO, these bfn_x, bfn_y, bfn_z are excessive. Having bfn_t is more than enough.

File: sycl/include/sycl/ext/intel/experimental/esimd/math.hpp
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/experimental/esimd/math.hpp:1733]: Surely, the enum defined below needs some doxy-gen comment.
The lines with `tparam` and 'param' looks a bit off here.

I would just write something like:L
```suggestion
/// This enum is used to encode all possible combinations of logical
/// operations performed on 3 input operand. It is used as a template
/// argument of bfn() function.
/// Example: d = bfn<~bfn::x & ~bfn::y & ~bfn::z>(s0, s1, s2);
```

File: sycl/include/sycl/ext/intel/experimental/esimd/math.hpp
Comment (v-klochkov) [inline sycl/include/sycl/ext/intel/experimental/esimd/math.hpp:1802]: for int64_t it is trivial for us to support it by using bitcast to int32_t elements.
For int8_t it should also be possible to support by
a) using bitcast to int16_t if N is even, bfn, bitcast to int16_t
b) extending with 1 element (to N+1 = even), bitcast to int16_t, bfn, extract N elements.

File: sycl/include/sycl/ext/intel/experimental/esimd/math.hpp
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/experimental/esimd/math.hpp:1783]: for int64_t it is trivial for us to support it by using bitcast to int32_t elements.
For int8_t it should also be possible to support by
a) if N - even: using bitcast to int16_t, bfn, bitcast to int16_t
b) if N - odd: extending with 1 element (to N+1 = even), bitcast to int16_t, bfn, extract N elements.

File: sycl/include/sycl/ext/intel/experimental/esimd/math.hpp
Comment (turinevgeny) [reply sycl/include/sycl/ext/intel/experimental/esimd/math.hpp]: > is it possible to put all the bfn_t stuff inside a stuct so it's encapsulated? or does it have to be seperate?
> 
> also is it possible to move it to some detail header?

I want to continue having bfn_t castable to uint8_t. Or you had something else in mind?
I looked at experimental detail header for math, it looks like it mostly contains `__esimd_*` intrinsics, I'm not sure it would belong there. Let me know what you think.

Comment (turinevgeny) [reply sycl/include/sycl/ext/intel/experimental/esimd/math.hpp]: > IMO, these bfn_x, bfn_y, bfn_z are excessive. Having bfn_t is more than enough.

Removed.

File: sycl/include/sycl/ext/intel/experimental/esimd/math.hpp
Comment (turinevgeny) [reply sycl/include/sycl/ext/intel/experimental/esimd/math.hpp:1775]: > just for my own interest where does the 2/4 byte requirement come from? does the underlying hardware instruction only support these?

Yes, it comes from HW.

Comment (turinevgeny) [reply sycl/include/sycl/ext/intel/experimental/esimd/math.hpp:1775]: > for int64_t it is trivial for us to support it by using bitcast to int32_t elements. For int8_t it should also be possible to support by a) if N - even: using bitcast to int16_t, bfn, bitcast to int16_t b) if N - odd: extending with 1 element (to N+1 = even), bitcast to int16_t, bfn, extract N elements.

Done.

File: sycl/test/esimd/intrins_trans.cpp
Comment (turinevgeny) [reply sycl/test/esimd/intrins_trans.cpp:310]: > sorry also for my own interest, the `i8 -1` argument is from the `0xff` template argument right?

That's right.

File: sycl/include/sycl/ext/intel/experimental/esimd/math.hpp
Comment (turinevgeny) [reply sycl/include/sycl/ext/intel/experimental/esimd/math.hpp]: Done. I did it without `constexpr if`s as some of implementations reuse others. Like `int8_t` with odd `N` would use `int8` with even `N`. Let me know what you think.

**Comment by v-klochkov** [comment]
@vmustya - You asked to show you the changes for BFN before merge. If you want to review the changes - it is a good moment. Thank you.

**Review by v-klochkov — APPROVED** [review]

**Review by sarnex — APPROVED** [review]

File: sycl/include/sycl/ext/intel/experimental/esimd/math.hpp
Comment (vmustya) [inline sycl/include/sycl/ext/intel/experimental/esimd/math.hpp:1786]: The target code may be more efficient, if you add a `N % 4 == 0` case with bitcast the args into i32

Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/experimental/esimd/math.hpp:1786]: @vmustya - can you give a bit more details. Do you say that BFN instruction for 2xi32 is faster than BFN for 4xi16?

Comment (vmustya) [reply sycl/include/sycl/ext/intel/experimental/esimd/math.hpp:1786]: Our hardware supports up to SIMD32 operations, so when you have `<128 x i8>` it would be more efficient to generate single `<32 x i32>` BFN instead of a couple of `<32 x i16>` instructions:
```LLVM
%x = call <32 x i32> @llvm.genx.bfn.v32i32(....) ; <- this is better then the following:
%x1 = call <32 x i16> @llvm.genx.bfn.v32i16(....)
%x2 = call <32 x i16> @llvm.genx.bfn.v32i16(....)
```

Comment (turinevgeny) [reply sycl/include/sycl/ext/intel/experimental/esimd/math.hpp:1788]: Done.

File: sycl/include/sycl/ext/intel/experimental/esimd/math.hpp
Comment (turinevgeny) [reply sycl/include/sycl/ext/intel/experimental/esimd/math.hpp]: Done.

Comment (turinevgeny) [reply sycl/include/sycl/ext/intel/experimental/esimd/math.hpp]: Done.

File: sycl/include/sycl/ext/intel/experimental/esimd/math.hpp
Comment (v-klochkov) [reply sycl/include/sycl/ext/intel/experimental/esimd/math.hpp:1789]: If use this approach/opt for bytes, then why not use for shorts?
I.e.:
```
template <bfn_t FuncControl, typename T, int N>
__ESIMD_API std::enable_if_t<std::is_integral_v<T>, __ESIMD_NS::simd<T, N>>
bfn(__ESIMD_NS::simd<T, N> src0, __ESIMD_NS::simd<T, N> src1,
    __ESIMD_NS::simd<T, N> src2) {
  if constexpr ((sizeof(T) == 8) || ((sizeof(T) == 1) && (N % 4 == 0)) ||
                ((sizeof(T) == 2) && (N % 2 == 0))) {
    // Bitcast Nx8-byte vectors to 2xN vectors of 4-byte integers.
    // Bitcast Nx1-byte vectors to N/4 vectors of 4-byte integers.
    // Bitcast Nx2-byte vectors to N/2 vectors of 4-byte integers.
    auto Result = __ESIMD_ENS::bfn<FuncControl>(
        src0.template bit_cast_view<int32_t>().read(),
        src1.template bit_cast_view<int32_t>().read(),
        src2.template bit_cast_view<int32_t>().read());
    return Result.template bit_cast_view<T>();
  } else if constexpr (sizeof(T) == 2 || sizeof(T) == 4) {
    constexpr uint8_t FC = static_cast<uint8_t>(FuncControl);
    return __esimd_bfn<FC, T, N>(src0.data(), src1.data(), src2.data());
  } else if constexpr (N % 2 == 0) {
    // Bitcast Nx1-byte vectors (N is even) to N/2 vectors of 4-byte integers.
    auto Result = __ESIMD_ENS::bfn<FuncControl>(
        src0.template bit_cast_view<int16_t>().read(),
        src1.template bit_cast_view<int16_t>().read(),
        src2.template bit_cast_view<int16_t>().read());
    return Result.template bit_cast_view<T>();
  } else {
    // Odd number of 1-byte elements.
    __ESIMD_NS::simd<T, N + 1> Src0, Src1, Src2;
    Src0.template select<N, 1>() = src0;
    Src1.template select<N, 1>() = src1;
    Src2.template select<N, 1>() = src2;
    auto Result = __ESIMD_ENS::bfn<FuncControl>(Src0, Src1, Src2);
    return Result.template select<N, 1>();
  }
}    
```

Comment (turinevgeny) [reply sycl/include/sycl/ext/intel/experimental/esimd/math.hpp:1791]: Done.

**Pull request merged**
